File: openswap-kotlin/lib/src/test/kotlin/org/openswap/ApiContractTest.kt
package org.openswap

import org.junit.jupiter.api.Test
import java.util.UUID
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.test.fail

/** Network-free contract tests for the generated Kotlin binding. */
class ApiContractTest {

    @Test
    fun takerMethodSurfaceIsComplete() {
        val expected = setOf(
            "backup",
            "displayOffer",
            "fetchAllMakers",
            "fetchOffers",
            "getBalances",
            "getNextExternalAddress",
            "getNextInternalAddresses",
            "getTransactions",
            "getWalletName",
            "listAllUtxoSpendInfo",
            "lockUnspendableUtxos",
            "pollMaker",
            "prepareOpenswap",
            "recoverActiveSwap",
            "removeMaker",
            "sendToAddress",
            "setupLogging",
            "startOpenswap",
            "syncAndSave",
            "syncOfferbookAndWait",
            "verifyDeniability",
        )
        // Kotlin value classes such as UInt add a generated "-<hash>" suffix to
        // JVM method names; compare the public Kotlin names represented by them.
        val actual = Taker::class.java.methods
            .mapTo(mutableSetOf()) { it.name.substringBefore('-') }

        assertEquals(emptySet(), expected - actual)

        val expectedGlobals = setOf(
            "createDefaultRpcConfig",
            "fetchMempoolFees",
            "isWalletEncrypted",
            "openswapFfiVersion",
            "restoreWalletGuiApp",
            "setupLogging",
        )
        val actualGlobals = Class.forName("org.openswap.OpenswapKt")
            .methods
            .mapTo(mutableSetOf()) { it.name }
        assertEquals(emptySet(), expectedGlobals - actualGlobals)
    }

    @Test
    fun nativeGlobalsAndTypedValidationErrors() {
        assertEquals("0.1.0", openswapFfiVersion())
        assertEquals(
            RpcConfig(
                url = "http://127.0.0.1:38332",
                username = "user",
                password = "password",
                walletName = "openswap_wallet",
            ),
            createDefaultRpcConfig(),
        )
        val missingWallet = "${System.getProperty("java.io.tmpdir")}/openswap-${UUID.randomUUID()}"
        assertFalse(isWalletEncrypted(missingWallet))

        val invalidBackend = BackendConfig(
            kind = "invalid",
            url = null,
            username = null,
            password = null,
            walletName = null,
            zmqAddr = null,
            socks5 = null,
            timeout = null,
            pollIntervalSecs = null,
            maxRetries = null,
        )
        val error = try {
            Taker.init(
                dataDir = null,
                walletFileName = null,
                rpcConfig = null,
                controlPort = null,
                torAuthPassword = null,
                zmqAddr = "tcp://127.0.0.1:28332",
                password = null,
                nostrRelays = emptyList(),
                backendConfig = invalidBackend,
                checkBlocklist = null,
            )
            fail("an unknown backend must be rejected before native resources are started")
        } catch (error: TakerException) {
            error
        }
        val general = assertIs<TakerException.General>(error)
        assertEquals("Invalid backend kind: invalid (expected rpc or electrum)", general.msg)
    }

    @Test
    fun scalarAndConfigurationRecordsPreserveEveryField() {
        assertEquals("bc1qexample", Address("bc1qexample").addr)
        assertEquals("P2TR", AddressType("P2TR").addrType)
        assertEquals(Long.MIN_VALUE, SignedAmountSats(Long.MIN_VALUE).sats)
        assertEquals("ab".repeat(32), Txid("ab".repeat(32)).value)
        assertEquals("0051ff", ScriptBuf("0051ff").hex)

        val keyBytes = ByteArray(33) { it.toByte() }
        val publicKey = PublicKey(compressed = true, inner = keyBytes)
        assertTrue(publicKey.compressed)
        assertContentEquals(keyBytes, publicKey.inner)

        val backend = BackendConfig(
            kind = "electrum",
            url = "ssl://electrum.example:50002",
            username = null,
            password = null,
            walletName = null,
            zmqAddr = null,
            socks5 = "127.0.0.1:9050",
            timeout = 120u.toUByte(),
            pollIntervalSecs = 15uL,
            maxRetries = 8u.toUByte(),
        )
        assertEquals("electrum", backend.kind)
        assertEquals("ssl://electrum.example:50002", backend.url)
        assertNull(backend.username)
        assertNull(backend.password)
        assertNull(backend.walletName)
        assertNull(backend.zmqAddr)
        assertEquals("127.0.0.1:9050", backend.socks5)
        assertEquals(120u.toUByte(), backend.timeout)
        assertEquals(15uL, backend.pollIntervalSecs)
        assertEquals(8u.toUByte(), backend.maxRetries)

        val balances = Balances(regular = 1, swap = 2, contract = 3, fidelity = 4, spendable = 3)
        assertEquals(listOf(1L, 2L, 3L, 4L, 3L), listOf(
            balances.regular,
            balances.swap,
            balances.contract,
            balances.fidelity,
            balances.spendable,
        ))
        val feeRates = FeeRates(12.5, 6.25, 1.0)
        assertEquals(listOf(12.5, 6.25, 1.0), listOf(
            feeRates.fastest,
            feeRates.standard,
            feeRates.economy,
        ))
        assertEquals("Blocks", LockTime("Blocks", 144u).lockType)
        assertEquals(144u, LockTime("Blocks", 144u).value)
        assertEquals("maker.onion:6102", MakerAddress("maker.onion:6102").address)
        val makerState = MakerState("Unresponsive", 7u.toUByte())
        assertEquals("Unresponsive", makerState.stateType)
        assertEquals(7u.toUByte(), makerState.retries)
        assertEquals("Unified", MakerProtocol("Unified").protocolType)
        val change = UtxoWithAddress(50_000, "bc1qchange")
        assertEquals(50_000L, change.amount)
        assertEquals("bc1qchange", change.address)
    }

    @Test
    fun transactionAndUtxoRecordsPreserveNestedOptionalData() {
        val txid = Txid("01".repeat(32))
        val info = WalletTxInfo(
            confirmations = -1,
            blockhash = "02".repeat(32),
            blockindex = 3u,
            blocktime = 1_700_000_000,
            blockheight = 250u,
            txid = txid,
            time = 1_700_000_001,
            timereceived = 1_700_000_002,
            bip125Replaceable = "Yes",
            walletConflicts = listOf(Txid("03".repeat(32))),
        )
        val detail = GetTransactionResultDetail(
            address = Address("bc1qtransaction"),
            category = "Send",
            amount = SignedAmountSats(-50_000),
            label = "payment",
            vout = 2u,
            fee = SignedAmountSats(-250),
            abandoned = false,
        )
        val transaction = ListTransactionResult(info, detail, trusted = true, comment = "memo")
        assertEquals(info, transaction.info)
        assertEquals(detail, transaction.detail)
        assertEquals(true, transaction.trusted)
        assertEquals("memo", transaction.comment)

        val unspent = ListUnspentResultEntry(
            txid = txid,
            vout = 4u,
            address = "bc1qutxo",
            label = "seed",
            scriptPubKey = ScriptBuf("0014"),
            amount = Amount(75_000),
            confirmations = 6u,
            redeemScript = ScriptBuf("51"),
            witnessScript = null,
            spendable = true,
            solvable = true,
            desc = "wpkh(...)",
            safe = false,
        )
        val spendInfo = UtxoSpendInfo(
            spendType = "FidelityBondCoin",
            path = "m/84'/1'/0'/0/1",
            multisigRedeemscript = null,
            inputValue = Amount(75_000),
            index = 9u,
        )
        val total = TotalUtxoInfo(unspent, spendInfo)
        assertEquals(unspent, total.listUnspentResultEntry)
        assertEquals(spendInfo, total.utxoSpendInfo)
    }

    @Test
    fun offerAndSwapRecordsPreserveTheCompleteNestedGraph() {
        val outpoint = OutPoint(Txid("04".repeat(32)), 5u)
        val publicKey = PublicKey(true, ByteArray(33) { 2 })
        val bond = FidelityBond(
            outpoint = outpoint,
            amount = Amount(100_000),
            lockTime = LockTime("Seconds", 500_000_000u),
            pubkey = publicKey,
            confHeight = 100u,
            certExpiry = null,
            isSpent = false,
        )
        val proof = FidelityProof(bond, ByteArray(32) { 5 }, ByteArray(64) { 6 })
        val offer = Offer(
            baseFee = -5,
            amountRelativeFeePct = 0.125,
            timeRelativeFeePct = 0.25,
            requiredConfirms = 2u,
            minimumLocktime = 48u.toUShort(),
            maxSize = 2_000_000,
            minSize = 50_000,
            tweakablePoint = publicKey,
            fidelity = proof,
        )
        val candidate = MakerOfferCandidate(
            address = MakerAddress("maker.onion:6102"),
            offer = offer,
            state = MakerState("Good", null),
            protocol = MakerProtocol("Taproot"),
        )
        val offerBook = OfferBook(listOf(candidate))
        assertEquals(candidate, offerBook.makers.single())
        assertEquals(-5L, offerBook.makers.single().offer?.baseFee)
        assertContentEquals(ByteArray(64) { 6 }, offerBook.makers.single().offer!!.fidelity.certSig)

        val params = SwapParams(
            protocol = "Taproot",
            sendAmount = 500_000uL,
            makerCount = 2u,
            txCount = 3u,
            requiredConfirms = 4u,
            manuallySelectedOutpoints = listOf(outpoint),
            preferredMakers = listOf("maker.onion:6102"),
            paymentAddress = null,
        )
        assertEquals("Taproot", params.protocol)
        assertEquals(500_000uL, params.sendAmount)
        assertEquals(2u, params.makerCount)
        assertEquals(3u, params.txCount)
        assertEquals(4u, params.requiredConfirms)
        assertEquals(listOf(outpoint), params.manuallySelectedOutpoints)
        assertEquals(listOf("maker.onion:6102"), params.preferredMakers)
        assertNull(params.paymentAddress)

        val fee = MakerFeeInfo(1u, "maker.onion:6102", 100.0, 200.0, 300.0, 600.0)
        val change = UtxoWithAddress(25_000, "bc1qchange")
        val swapOutput = UtxoWithAddress(475_000, "bc1qswap")
        val outgoingUtxo = ReportUtxo("bc1qoutgoing", 510_000)
        val incomingUtxo = ReportUtxo("bc1qincoming", 475_000)
        val report = SwapReport(
            swapId = "swap-1",
            role = "Taker",
            status = "SUCCESS",
            swapDurationSeconds = 12.5,
            startTimestamp = 1_700_000_000,
            endTimestamp = 1_700_000_013,
            network = "regtest",
            errorMessage = null,
            incomingAmount = 500_000,
            outgoingAmount = 510_000,
            feePaid = -10_000,
            outgoingUtxos = listOf(outgoingUtxo),
            incomingUtxos = listOf(incomingUtxo),
            fundingTxids = listOf(listOf("08".repeat(32)), listOf("09".repeat(32))),
            makersCount = 2u,
            makerAddresses = listOf("maker-1", "maker-2"),
            totalMakerFees = 8_000,
            miningFee = 2_000,
            feePercentage = 2.0,
            makerFeeInfo = listOf(fee),
            inputUtxos = listOf(510_000),
            outputChangeAmounts = listOf(25_000),
            outputSwapAmounts = listOf(475_000),
            outputChangeUtxos = listOf(change),
            outputSwapUtxos = listOf(swapOutput),
        )
        assertEquals("swap-1", report.swapId)
        assertEquals("Taker", report.role)
        assertEquals("SUCCESS", report.status)
        assertEquals(12.5, report.swapDurationSeconds)
        assertEquals(1_700_000_000L, report.startTimestamp)
        assertEquals(1_700_000_013L, report.endTimestamp)
        assertEquals("regtest", report.network)
        assertNull(report.errorMessage)
        assertEquals(500_000L, report.incomingAmount)
        assertEquals(510_000L, report.outgoingAmount)
        assertEquals(-10_000L, report.feePaid)
        assertEquals(listOf(outgoingUtxo), report.outgoingUtxos)
        assertEquals(listOf(incomingUtxo), report.incomingUtxos)
        assertEquals(2, report.fundingTxids.size)
        assertEquals(2u, report.makersCount)
        assertEquals(listOf("maker-1", "maker-2"), report.makerAddresses)
        assertEquals(8_000L, report.totalMakerFees)
        assertEquals(2_000L, report.miningFee)
        assertEquals(2.0, report.feePercentage)
        assertEquals(listOf(fee), report.makerFeeInfo)
        assertEquals(listOf(510_000L), report.inputUtxos)
        assertEquals(listOf(25_000L), report.outputChangeAmounts)
        assertEquals(listOf(475_000L), report.outputSwapAmounts)
        assertEquals(listOf(change), report.outputChangeUtxos)
        assertEquals(listOf(swapOutput), report.outputSwapUtxos)
    }

    @Test
    fun enumAndErrorVariantsRemainDistinctAndKeepMessages() {
        assertEquals(3, TakerBehavior.entries.size)
        assertEquals(
            setOf(
                TakerBehavior.NORMAL,
                TakerBehavior.DROP_CONNECTION_AFTER_FULL_SETUP,
                TakerBehavior.BROADCAST_CONTRACT_AFTER_FULL_SETUP,
            ),
            TakerBehavior.entries.toSet(),
        )
        listOf(
            TakerException.Wallet("reason").msg,
            TakerException.Protocol("reason").msg,
            TakerException.Network("reason").msg,
            TakerException.General("reason").msg,
            TakerException.Io("reason").msg,
        ).forEach { assertEquals("reason", it) }
    }
}

File: openswap-csharp/Openswap.Tests/ApiContractTest.cs
using Openswap.Native;
using Xunit;

namespace Openswap.Tests;

/// <summary>Network-free contracts for values and errors crossing the .NET FFI boundary.</summary>
public class ApiContractTest
{
    [Fact]
    public void TakerMethodSurfaceIsComplete()
    {
        string[] expected = [
            "Backup", "DisplayOffer", "FetchAllMakers", "FetchOffers", "GetBalances",
            "GetNextExternalAddress", "GetNextInternalAddresses", "GetTransactions",
            "GetWalletName", "ListAllUtxoSpendInfo", "LockUnspendableUtxos", "PollMaker",
            "PrepareOpenswap", "RecoverActiveSwap", "RemoveMaker", "SendToAddress",
            "SetupLogging", "StartOpenswap", "SyncAndSave", "SyncOfferbookAndWait",
            "VerifyDeniability",
        ];
        var actual = typeof(Taker).GetMethods()
            .Select(method => method.Name)
            .ToHashSet(StringComparer.Ordinal);

        Assert.Empty(expected.Where(method => !actual.Contains(method)));

        string[] expectedGlobals = [
            "CreateDefaultRpcConfig", "FetchMempoolFees", "IsWalletEncrypted",
            "OpenswapFfiVersion", "RestoreWalletGuiApp", "SetupLogging",
        ];
        var actualGlobals = typeof(global::Openswap.Native.Openswap).GetMethods()
            .Select(method => method.Name)
            .ToHashSet(StringComparer.Ordinal);
        Assert.Empty(expectedGlobals.Where(method => !actualGlobals.Contains(method)));
    }

    [Fact]
    public void NativeDefaultsAndValidationErrorsAreStable()
    {
        Assert.Equal("0.1.0", OpenswapClient.NativeVersion);
        Assert.Equal(
            new RpcConfig(
                Url: "http://127.0.0.1:38332",
                Username: "user",
                Password: "password",
                WalletName: "openswap_wallet"),
            OpenswapClient.DefaultRpcConfig());
        Assert.False(global::Openswap.Native.Openswap.IsWalletEncrypted(
            Path.Combine(Path.GetTempPath(), $"openswap-{Guid.NewGuid():N}")));

        var invalidBackend = new BackendConfig(
            Kind: "invalid", Url: null, Username: null, Password: null,
            WalletName: null, ZmqAddr: null, Socks5: null, Timeout: null,
            PollIntervalSecs: null, MaxRetries: null);
        var error = Assert.Throws<TakerException.General>(() => Taker.Init(
            null, null, null, null, null, "tcp://127.0.0.1:28332", null,
            Array.Empty<string>(), invalidBackend));
        Assert.Equal(
            "Invalid backend kind: invalid (expected rpc or electrum)",
            error.msg);
    }

    [Fact]
    public void ScalarConfigurationAndWalletRecordsPreserveEveryField()
    {
        Assert.Equal("bc1qexample", new Address("bc1qexample").Addr);
        Assert.Equal("P2TR", new AddressType("P2TR").AddrType);
        Assert.Equal(long.MinValue, new SignedAmountSats(long.MinValue).Sats);
        Assert.Equal("0051ff", new ScriptBuf("0051ff").Hex);

        var txid = new Txid(string.Concat(Enumerable.Repeat("01", 32)));
        var keyBytes = Enumerable.Range(0, 33).Select(i => (byte)i).ToArray();
        var publicKey = new PublicKey(Compressed: true, Inner: keyBytes);
        Assert.True(publicKey.Compressed);
        Assert.Equal(keyBytes, publicKey.Inner);

        var backend = new BackendConfig(
            Kind: "electrum",
            Url: "ssl://electrum.example:50002",
            Username: null,
            Password: null,
            WalletName: null,
            ZmqAddr: null,
            Socks5: "127.0.0.1:9050",
            Timeout: 120,
            PollIntervalSecs: 15,
            MaxRetries: 8);
        Assert.Equal("electrum", backend.Kind);
        Assert.Equal("ssl://electrum.example:50002", backend.Url);
        Assert.Equal("127.0.0.1:9050", backend.Socks5);
        Assert.Equal((byte?)120, backend.Timeout);
        Assert.Equal((ulong?)15, backend.PollIntervalSecs);
        Assert.Equal((byte?)8, backend.MaxRetries);

        var info = new WalletTxInfo(
            Confirmations: -1,
            Blockhash: string.Concat(Enumerable.Repeat("02", 32)),
            Blockindex: 3,
            Blocktime: 1_700_000_000,
            Blockheight: 250,
            Txid: txid,
            Time: 1_700_000_001,
            Timereceived: 1_700_000_002,
            Bip125Replaceable: "Yes",
            WalletConflicts: [new Txid(string.Concat(Enumerable.Repeat("03", 32)))]);
        var detail = new GetTransactionResultDetail(
            Address: new Address("bc1qtransaction"),
            Category: "Send",
            Amount: new SignedAmountSats(-50_000),
            Label: "payment",
            Vout: 2,
            Fee: new SignedAmountSats(-250),
            Abandoned: false);
        var transaction = new ListTransactionResult(info, detail, true, "memo");
        Assert.Equal(info, transaction.Info);
        Assert.Equal(detail, transaction.Detail);
        Assert.True(transaction.Trusted is true);
        Assert.Equal("memo", transaction.Comment);

        var unspent = new ListUnspentResultEntry(
            Txid: txid,
            Vout: 4,
            Address: "bc1qutxo",
            Label: "seed",
            ScriptPubKey: new ScriptBuf("0014"),
            Amount: new Amount(75_000),
            Confirmations: 6,
            RedeemScript: new ScriptBuf("51"),
            WitnessScript: null,
            Spendable: true,
            Solvable: true,
            Desc: "wpkh(...)",
            Safe: false);
        var spendInfo = new UtxoSpendInfo(
            SpendType: "FidelityBondCoin",
            Path: "m/84'/1'/0'/0/1",
            MultisigRedeemscript: null,
            InputValue: new Amount(75_000),
            Index: 9);
        var total = new TotalUtxoInfo(unspent, spendInfo);
        Assert.Equal(unspent, total.ListUnspentResultEntry);
        Assert.Equal(spendInfo, total.UtxoSpendInfo);

        var balances = new Balances(1, 2, 3, 4, 3);
        Assert.Equal(
            new long[] { 1, 2, 3, 4, 3 },
            new[] { balances.Regular, balances.Swap, balances.Contract, balances.Fidelity, balances.Spendable });
        var lockTime = new LockTime("Blocks", 144);
        Assert.Equal("Blocks", lockTime.LockType);
        Assert.Equal((uint)144, lockTime.Value);
        Assert.Equal("maker.onion:6102", new MakerAddress("maker.onion:6102").Address);
        var unavailable = new UnavailableState("NoOfferResponse", 100, 200, 7);
        var makerState = new MakerState("Unavailable", unavailable, null);
        Assert.Equal("Unavailable", makerState.StateType);
        Assert.Equal("NoOfferResponse", makerState.Unavailable?.Reason);
        Assert.Equal((ulong?)100, makerState.Unavailable?.SinceTs);
        Assert.Equal((ulong?)200, makerState.Unavailable?.LastAttemptTs);
        Assert.Equal((uint?)7, makerState.Unavailable?.Attempts);
        var banned = new MakerState(
            "Banned", null, new BanRecord("InvalidFidelityProof", 300));
        Assert.Equal("InvalidFidelityProof", banned.Ban?.Reason);
        Assert.Equal((ulong?)300, banned.Ban?.RecordedAtTs);
        Assert.Equal("Unified", new MakerProtocol("Unified").ProtocolType);
    }

    [Fact]
    public void OfferAndSwapRecordsPreserveTheCompleteNestedGraph()
    {
        var outpoint = new OutPoint(
            new Txid(string.Concat(Enumerable.Repeat("04", 32))), 5);
        var publicKey = new PublicKey(true, Enumerable.Repeat((byte)2, 33).ToArray());
        var bond = new FidelityBond(
            Outpoint: outpoint,
            Amount: new Amount(100_000),
            LockTime: new LockTime("Seconds", 500_000_000),
            Pubkey: publicKey,
            ConfHeight: 100,
            CertExpiry: null,
            IsSpent: false);
        var proof = new FidelityProof(
            Bond: bond,
            CertHash: Enumerable.Repeat((byte)5, 32).ToArray(),
            CertSig: Enumerable.Repeat((byte)6, 64).ToArray());
        var offer = new Offer(
            BaseFee: -5,
            AmountRelativeFeePct: 0.125,
            TimeRelativeFeePct: 0.25,
            RequiredConfirms: 2,
            MinimumLocktime: 48,
            MaxSize: 2_000_000,
            MinSize: 50_000,
            TweakablePoint: publicKey,
            Fidelity: proof);
        var candidate = new MakerOfferCandidate(
            Address: new MakerAddress("maker.onion:6102"),
            Offer: offer,
            State: new MakerState("Good", null, null),
            Protocol: new MakerProtocol("Taproot"));
        Assert.Equal(candidate, new OfferBook([candidate]).Makers.Single());
        Assert.Equal(Enumerable.Repeat((byte)6, 64), candidate.Offer!.Fidelity.CertSig);

        var parameters = new SwapParams(
            Protocol: "Taproot",
            SendAmount: 500_000,
            MakerCount: 2,
            TxCount: 3,
            RequiredConfirms: 4,
            ManuallySelectedOutpoints: [outpoint],
            PreferredMakers: ["maker.onion:6102"],
            PaymentAddress: null,
            MaxInputBudget: 5,
            Feerate: 6);
        Assert.Equal("Taproot", parameters.Protocol);
        Assert.Equal((ulong)500_000, parameters.SendAmount);
        Assert.Equal((uint)2, parameters.MakerCount);
        Assert.Equal((uint?)3, parameters.TxCount);
        Assert.Equal((uint?)4, parameters.RequiredConfirms);
        Assert.Equal(outpoint, parameters.ManuallySelectedOutpoints!.Single());
        Assert.Equal("maker.onion:6102", parameters.PreferredMakers!.Single());
        Assert.Null(parameters.PaymentAddress);
        Assert.Equal((uint?)5, parameters.MaxInputBudget);
        Assert.Equal((ulong?)6, parameters.Feerate);

        var fee = new MakerFeeInfo(1, "maker.onion:6102", 100, 200, 300, 600);
        var change = new UtxoWithAddress(25_000, "bc1qchange");
        var swapOutput = new UtxoWithAddress(475_000, "bc1qswap");
        var outgoingUtxo = new ReportUtxo("bc1qoutgoing", 510_000);
        var incomingUtxo = new ReportUtxo("bc1qincoming", 475_000);
        var report = new SwapReport(
            SwapId: "swap-1",
            Role: "Taker",
            Status: "SUCCESS",
            SwapDurationSeconds: 12.5,
            StartTimestamp: 1_700_000_000,
            EndTimestamp: 1_700_000_013,
            Network: "regtest",
            ErrorMessage: null,
            IncomingAmount: 500_000,
            OutgoingAmount: 510_000,
            FeePaid: -10_000,
            OutgoingUtxos: [outgoingUtxo],
            IncomingUtxos: [incomingUtxo],
            FundingTxids: [
                [string.Concat(Enumerable.Repeat("08", 32))],
                [string.Concat(Enumerable.Repeat("09", 32))],
            ],
            MakersCount: 2,
            MakerAddresses: ["maker-1", "maker-2"],
            TotalMakerFees: 8_000,
            MiningFee: 2_000,
            FeePercentage: 2,
            MakerFeeInfo: [fee],
            InputUtxos: [510_000],
            OutputChangeAmounts: [25_000],
            OutputSwapAmounts: [475_000],
            OutputChangeUtxos: [change],
            OutputSwapUtxos: [swapOutput]);
        Assert.Equal("swap-1", report.SwapId);
        Assert.Equal(-10_000, report.FeePaid);
        Assert.Equal(outgoingUtxo, report.OutgoingUtxos.Single());
        Assert.Equal(incomingUtxo, report.IncomingUtxos.Single());
        Assert.Equal(2, report.FundingTxids.Length);
        Assert.Equal((uint?)2, report.MakersCount);
        Assert.Equal(fee, report.MakerFeeInfo.Single());
        Assert.Equal(change, report.OutputChangeUtxos.Single());
        Assert.Equal(swapOutput, report.OutputSwapUtxos.Single());
    }

    [Fact]
    public void EnumAndErrorVariantsRemainDistinctAndKeepMessages()
    {
        Assert.Equal(3, Enum.GetValues<TakerBehavior>().Length);
        Assert.Equal("reason", new TakerException.Wallet("reason").msg);
        Assert.Equal("reason", new TakerException.Protocol("reason").msg);
        Assert.Equal("reason", new TakerException.Network("reason").msg);
        Assert.Equal("reason", new TakerException.General("reason").msg);
        Assert.Equal("reason", new TakerException.Io("reason").msg);
    }
}
